server/src/main/java/org/opensearch/telemetry/tracing/ThreadContextBasedTracerContextStorage.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.telemetry.tracing;
+
+import org.opensearch.common.util.concurrent.ThreadContext;
+
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Core's ThreadContext based TracerContextStorage implementation
+ */
+public class ThreadContextBasedTracerContextStorage implements TracerContextStorage<String, Span> {
+
+    private final ThreadContext threadContext;
+
+    private final TracingTelemetry tracingTelemetry;
+
+    public ThreadContextBasedTracerContextStorage(ThreadContext threadContext, TracingTelemetry tracingTelemetry) {
+        this.threadContext = threadContext;
+        this.tracingTelemetry = tracingTelemetry;
+    }
+
+    @Override
+    public Span get(String key) {
+        return getCurrentSpan(key);
+    }
+
+    @Override
+    public void put(String key, Span span) {
+        if (span == null) {
+            return;
+        }
+        AtomicReference<Span> currentSpanRef = threadContext.getTransient(key);
```

Comment:
What is the purpose to use `AtomicReference` here? This is by definition per-thread context storage, no races are expected.

Reply:
We are actually using `AtomicReference` as a wrapper to store the span in transient headers. Transient headers does not allow to mutate its key value once it is set. Hence as a workaround we updating the value instead(which is atomic reference).

Reply:
Oh I see, in case the current span changes in scope of the current thread, I think we should introduce lightweight `SpanReference` class in this case, using `AtomicReference` is too much.

Reply:
Done